test/jdk/java/util/concurrent/tck/ForkJoinPool20Test.java

```diff
@@ -216,4 +238,416 @@ static ForkJoinWorkerThread submitBusyTask(ForkJoinPool pool) throws Exception {
+                    return Boolean.TRUE;
+                } catch (InterruptedException shutdownNowDeliberatelyIgnored) {}
+            }};
+
+        try (PoolCleaner cleaner = cleaner(p, done)) {
+            for (int i = p.getParallelism(); i--> 0; ) {
+                switch (rnd.nextInt(4)) {
+                case 0: p.execute(r); break;
+                case 1: assertFalse(p.submit(r).isDone()); break;
+                case 2: assertFalse(p.submit(r, Boolean.TRUE).isDone()); break;
+                case 3: assertFalse(p.submit(c).isDone()); break;
+                }
+            }
+
+            await(threadsStarted);
+            p.shutdown();
+            done.countDown();   // release blocking tasks
+            assertTrue(p.awaitTermination(LONG_DELAY_MS, MILLISECONDS));
+
+        }
+    }
+
+    /**
+     * A fixed delay task with overflowing period should not prevent a
+     * one-shot task from executing.
+     * https://bugs.openjdk.org/browse/JDK-8051859
+     */
+    @SuppressWarnings("FutureReturnValueIgnored")
+    public void testScheduleWithFixedDelay_overflow() throws Exception {
+        final CountDownLatch delayedDone = new CountDownLatch(1);
+        final CountDownLatch immediateDone = new CountDownLatch(1);
+        final ForkJoinPool p = new ForkJoinPool(2);
+        try (PoolCleaner cleaner = cleaner(p)) {
+            final Runnable delayed = () -> {
+                delayedDone.countDown();
+                p.submit(() -> immediateDone.countDown());
+            };
+            p.scheduleWithFixedDelay(delayed, 0L, Long.MAX_VALUE, SECONDS);
+            await(delayedDone);
+            await(immediateDone);
+        }
+    }
+
+    /**
+     * shutdownNow cancels tasks that were not run
+     */
+    public void testShutdownNow_delayedTasks() throws InterruptedException {
+        final ForkJoinPool p = new ForkJoinPool(2);
+        List<ScheduledFuture<?>> tasks = new ArrayList<>();
+        for (int i = 0; i < 3; i++) {
+            Runnable r = new NoOpRunnable();
+            tasks.add(p.schedule(r, 9, SECONDS));
+            tasks.add(p.scheduleAtFixedRate(r, 9, 9, SECONDS));
+            tasks.add(p.scheduleWithFixedDelay(r, 9, 9, SECONDS));
+        }
+        p.shutdownNow();
+        assertTrue(p.awaitTermination(LONG_DELAY_MS, MILLISECONDS));
+        for (ScheduledFuture<?> task : tasks) {
+            assertTrue(task.isDone());
+        }
+        assertTrue(p.isTerminated());
+    }
+
+    /**
+     * submitWithTimeout cancels task after timeout
+     */
+    public void testSubmitWithTimeoutCancels() throws InterruptedException {
+        final ForkJoinPool p = ForkJoinPool.commonPool();
+        Callable<Boolean> c = new Callable<Boolean>() {
+            public Boolean call() throws Exception {
+                Thread.sleep(LONGER_DELAY_MS); return Boolean.TRUE; }};
+        ForkJoinTask<?> task = p.submitWithTimeout(c, 1, NANOSECONDS, null);
+        Thread.sleep(timeoutMillis());
```

Comment:
Do we need to sleep here or just get the result with a timeout expecting a CancellationException then asserting isCancelled == true?

Reply:
This checks: The task is not run and will eventually show up as cancelled. So as is done in other juc-tck test, we sleep a little for the "eventually" part. This and many similar cases can rarely misfire, in which case timeoutMillis is increased as a check by harness.